src/providers/bibtexformater.ts

```diff
@@ -67,56 +96,67 @@ export class BibtexFormater {
+                        entryLocations[i].start.character + columnOffset,
+                        entryLocations[i].start.line + lineDelta + (sortedEntryLocations[i].end.line - sortedEntryLocations[i].start.line) + lineOffset,
                         entryLocations[i].end.character
                     )
-                    diags.push(new vscode.Diagnostic(
+                    this.diags.push(new vscode.Diagnostic(
                         highlightRange,
                         `Duplicate entry "${entries[i].internalKey}".`,
                         vscode.DiagnosticSeverity.Warning
                     ))
                 } else { // 'Comment Duplicates'
                     // Log duplicate entry since we aren't highlighting it
                     this.extension.logger.addLogMessage(
-                        `BibTeX-format: Duplicate entry "${entries[i].internalKey}" at line ${entryLocations[i].start.line + lineDelta + 1}.`)
+                        `BibTeX-format: Duplicate entry "${entries[i].internalKey}" at line ${entryLocations[i].start.line + lineDelta + 1 + lineOffset}.`)
                     text = text.replace(/@/,'')
                 }
             }
 
             // Put text from entry[i] into (sorted)location[i]
-            edit.replace(uri, entryLocations[i], text)
+            edits.push(new vscode.TextEdit(new vscode.Range(entryLocations[i].start.translate(range?.start.line, range?.start.character), entryLocations[i].end.translate(range?.start.line)), text))
 
             // We need to figure out the line changes in order to highlight properly
             lineDelta += (sortedEntryLocations[i].end.line - sortedEntryLocations[i].start.line) - (entryLocations[i].end.line - entryLocations[i].start.line)
         }
+        this.extension.logger.addLogMessage('Formatted ' + document.fileName)
+        return edits
+    }
+}
 
-        vscode.workspace.applyEdit(edit).then(success => {
-            if (success) {
-                this.duplicatesDiagnostics.set(uri, diags)
-                const t1 = performance.now()
-                this.extension.logger.addLogMessage(`BibTeX action successful. Took ${t1 - t0} ms.`)
-            } else {
-                this.extension.logger.showErrorMessage('Something went wrong while processing the bibliography.')
-            }
-        })
+export class BibtexFormatterProvider implements vscode.DocumentFormattingEditProvider, vscode.DocumentRangeFormattingEditProvider {
+    private formatter: BibtexFormatter
+    extension: Extension
+
+    constructor(extension: Extension) {
+        this.extension = extension
+        this.formatter = new BibtexFormatter(extension)
+    }
+
+    public provideDocumentFormattingEdits(document: vscode.TextDocument, _options: vscode.FormattingOptions, _token: vscode.CancellationToken): vscode.ProviderResult<vscode.TextEdit[]> {
+        return this.formatter.formatDocument(document, true, true)
```

Comment:
I think sorting bib items always goes too far. It should be configurable.

Reply:
I am not sure to understand exactly.
Do you suggest to let the user configure whether he wants to sort the content or just align it?

Reply:
Yes. The current formatter, `latexindent.pl`, have not sorted items. Just aligning the content by default, not sorting, would be what users expect for a formatter by default.

Reply:
We should add to log messages why formatting starts. VS Code executed it as the formatter, or the user executed the command.